Origin: https://github.com/php/php-src/commit/ed70b1ea43a9b7ffa2f53b3e5d6ba403f37ae81c
Advisory: GHSA-www2-q4fc-65wf
Imported-by: backported to 7.2 by hand (Z_PARAM_PATH for getservbyname() name; 7.2 warning-style ZPP errors in tests; existing ZPP-message tests updated)

From 3fce1f6ef2fe1939c77ca595190c24fef3070f46 Mon Sep 17 00:00:00 2001
From: Niels Dossche <7771979+nielsdos@users.noreply.github.com>
Date: Sat, 6 Sep 2025 21:55:13 +0200
Subject: [PATCH] Fix GHSA-www2-q4fc-65wf

Backport of php-src ed70b1ea43a9b7ffa2f53b3e5d6ba403f37ae81c (PHP-8.1) to PHP 7.2:
- getservbyname() name is a char* in 7.2: Z_PARAM_PATH instead of Z_PARAM_PATH_STR
- dns_win32.c: "s" -> "p" in the existing 7.2 format strings
- test: 7.2 ZPP failures are a warning + NULL (false for inet_pton) instead of ValueError
- existing tests updated: wrong-type warnings now say "to be a valid path" instead of "to be string"
---
 ext/standard/basic_functions.c                | 12 +++---
 ext/standard/dns.c                            |  6 +--
 ext/standard/dns_win32.c                      |  6 +--
 .../getservbyname_variation1.phpt             |  8 ++--
 .../getservbyname_variation13.phpt            |  2 +-
 .../getservbyname_variation6.phpt             |  2 +-
 .../getservbyname_variation8.phpt             |  8 ++--
 .../tests/network/ghsa-www2-q4fc-65wf.phpt    | 43 +++++++++++++++++++
 ext/standard/tests/network/ip.phpt            |  2 +-
 .../tests/network/ip2long_variation1.phpt     | 12 +++---
 ext/standard/tests/network/ip_x86_64.phpt     |  2 +-
 11 files changed, 73 insertions(+), 30 deletions(-)
 create mode 100644 ext/standard/tests/network/ghsa-www2-q4fc-65wf.phpt

diff --git a/ext/standard/basic_functions.c b/ext/standard/basic_functions.c
index 52a67c9b..5d866feb 100644
--- a/ext/standard/basic_functions.c
+++ b/ext/standard/basic_functions.c
@@ -3939,7 +3939,7 @@ PHP_NAMED_FUNCTION(php_inet_pton)
 	char buffer[17];
 
 	ZEND_PARSE_PARAMETERS_START(1, 1)
-		Z_PARAM_STRING(address, address_len)
+		Z_PARAM_PATH(address, address_len)
 	ZEND_PARSE_PARAMETERS_END_EX(RETURN_FALSE);
 
 	memset(buffer, 0, sizeof(buffer));
@@ -3979,7 +3979,7 @@ PHP_FUNCTION(ip2long)
 #endif
 
 	ZEND_PARSE_PARAMETERS_START(1, 1)
-		Z_PARAM_STRING(addr, addr_len)
+		Z_PARAM_PATH(addr, addr_len)
 	ZEND_PARSE_PARAMETERS_END();
 
 #ifdef HAVE_INET_PTON
@@ -5660,8 +5660,8 @@ PHP_FUNCTION(getservbyname)
 	struct servent *serv;
 
 	ZEND_PARSE_PARAMETERS_START(2, 2)
-		Z_PARAM_STRING(name, name_len)
-		Z_PARAM_STRING(proto, proto_len)
+		Z_PARAM_PATH(name, name_len)
+		Z_PARAM_PATH(proto, proto_len)
 	ZEND_PARSE_PARAMETERS_END();
 
 
@@ -5705,7 +5705,7 @@ PHP_FUNCTION(getservbyport)
 
 	ZEND_PARSE_PARAMETERS_START(2, 2)
 		Z_PARAM_LONG(port)
-		Z_PARAM_STRING(proto, proto_len)
+		Z_PARAM_PATH(proto, proto_len)
 	ZEND_PARSE_PARAMETERS_END();
 
 	serv = getservbyport(htons((unsigned short) port), proto);
@@ -5729,7 +5729,7 @@ PHP_FUNCTION(getprotobyname)
 	struct protoent *ent;
 
 	ZEND_PARSE_PARAMETERS_START(1, 1)
-		Z_PARAM_STRING(name, name_len)
+		Z_PARAM_PATH(name, name_len)
 	ZEND_PARSE_PARAMETERS_END();
 
 	ent = getprotobyname(name);
diff --git a/ext/standard/dns.c b/ext/standard/dns.c
index 7d044223..f940f4d4 100644
--- a/ext/standard/dns.c
+++ b/ext/standard/dns.c
@@ -368,7 +368,7 @@ PHP_FUNCTION(dns_check_record)
 #endif
 
 	ZEND_PARSE_PARAMETERS_START(1, 2)
-		Z_PARAM_STRING(hostname, hostname_len)
+		Z_PARAM_PATH(hostname, hostname_len)
 		Z_PARAM_OPTIONAL
 		Z_PARAM_STRING(rectype, rectype_len)
 	ZEND_PARSE_PARAMETERS_END();
@@ -815,7 +815,7 @@ PHP_FUNCTION(dns_get_record)
 	zend_bool raw = 0;
 
 	ZEND_PARSE_PARAMETERS_START(1, 5)
-		Z_PARAM_STRING(hostname, hostname_len)
+		Z_PARAM_PATH(hostname, hostname_len)
 		Z_PARAM_OPTIONAL
 		Z_PARAM_LONG(type_param)
 		Z_PARAM_ZVAL_DEREF_EX(authns, 1, 0)
@@ -1050,7 +1050,7 @@ PHP_FUNCTION(dns_get_mx)
 #endif
 
 	ZEND_PARSE_PARAMETERS_START(2, 3)
-		Z_PARAM_STRING(hostname, hostname_len)
+		Z_PARAM_PATH(hostname, hostname_len)
 		Z_PARAM_ZVAL_DEREF(mx_list)
 		Z_PARAM_OPTIONAL
 		Z_PARAM_ZVAL_DEREF(weight_list)
diff --git a/ext/standard/dns_win32.c b/ext/standard/dns_win32.c
index 4bca7128..defc3aec 100644
--- a/ext/standard/dns_win32.c
+++ b/ext/standard/dns_win32.c
@@ -50,7 +50,7 @@ PHP_FUNCTION(dns_get_mx) /* {{{ */
 	DNS_STATUS      status;                 /* Return value of DnsQuery_A() function */
 	PDNS_RECORD     pResult, pRec;          /* Pointer to DNS_RECORD structure */
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "sz/|z/", &hostname, &hostname_len, &mx_list, &weight_list) == FAILURE) {
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "pz/|z/", &hostname, &hostname_len, &mx_list, &weight_list) == FAILURE) {
 		return;
 	}
 
@@ -99,7 +99,7 @@ PHP_FUNCTION(dns_check_record)
 	DNS_STATUS      status;                 /* Return value of DnsQuery_A() function */
 	PDNS_RECORD     pResult;          /* Pointer to DNS_RECORD structure */
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s|s", &hostname, &hostname_len, &rectype, &rectype_len) == FAILURE) {
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "p|s", &hostname, &hostname_len, &rectype, &rectype_len) == FAILURE) {
 		return;
 	}
 
@@ -352,7 +352,7 @@ PHP_FUNCTION(dns_get_record)
 	int type, type_to_fetch, first_query = 1, store_results = 1;
 	zend_bool raw = 0;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s|lz/!z/!b",
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "p|lz/!z/!b",
 			&hostname, &hostname_len, &type_param, &authns, &addtl, &raw) == FAILURE) {
 		return;
 	}
diff --git a/ext/standard/tests/general_functions/getservbyname_variation1.phpt b/ext/standard/tests/general_functions/getservbyname_variation1.phpt
index 7dd01aab..ac7d9fe0 100644
--- a/ext/standard/tests/general_functions/getservbyname_variation1.phpt
+++ b/ext/standard/tests/general_functions/getservbyname_variation1.phpt
@@ -27,14 +27,14 @@ foreach ( $variation_array as $var ) {
 --EXPECTF--
 *** Test substituting argument 1 with array values ***
 
-Warning: getservbyname() expects parameter 1 to be string, array given in %s on line %d
+Warning: getservbyname() expects parameter 1 to be a valid path, array given in %s on line %d
 NULL
 
-Warning: getservbyname() expects parameter 1 to be string, array given in %s on line %d
+Warning: getservbyname() expects parameter 1 to be a valid path, array given in %s on line %d
 NULL
 
-Warning: getservbyname() expects parameter 1 to be string, array given in %s on line %d
+Warning: getservbyname() expects parameter 1 to be a valid path, array given in %s on line %d
 NULL
 
-Warning: getservbyname() expects parameter 1 to be string, array given in %s on line %d
+Warning: getservbyname() expects parameter 1 to be a valid path, array given in %s on line %d
 NULL
diff --git a/ext/standard/tests/general_functions/getservbyname_variation13.phpt b/ext/standard/tests/general_functions/getservbyname_variation13.phpt
index 8dad8cda..b556d464 100644
--- a/ext/standard/tests/general_functions/getservbyname_variation13.phpt
+++ b/ext/standard/tests/general_functions/getservbyname_variation13.phpt
@@ -34,5 +34,5 @@ foreach ( $variation_array as $var ) {
 *** Test substituting argument 2 with object values ***
 bool(false)
 
-Warning: getservbyname() expects parameter 2 to be string, object given in %s.php on line %d
+Warning: getservbyname() expects parameter 2 to be a valid path, object given in %s.php on line %d
 NULL
diff --git a/ext/standard/tests/general_functions/getservbyname_variation6.phpt b/ext/standard/tests/general_functions/getservbyname_variation6.phpt
index 0dfafa62..e184c1a1 100644
--- a/ext/standard/tests/general_functions/getservbyname_variation6.phpt
+++ b/ext/standard/tests/general_functions/getservbyname_variation6.phpt
@@ -34,5 +34,5 @@ foreach ( $variation_array as $var ) {
 *** Test substituting argument 1 with object values ***
 bool(false)
 
-Warning: getservbyname() expects parameter 1 to be string, object given in %s.php on line %d
+Warning: getservbyname() expects parameter 1 to be a valid path, object given in %s.php on line %d
 NULL
diff --git a/ext/standard/tests/general_functions/getservbyname_variation8.phpt b/ext/standard/tests/general_functions/getservbyname_variation8.phpt
index 69d1d77b..78d7d0b7 100644
--- a/ext/standard/tests/general_functions/getservbyname_variation8.phpt
+++ b/ext/standard/tests/general_functions/getservbyname_variation8.phpt
@@ -27,14 +27,14 @@ foreach ( $variation_array as $var ) {
 --EXPECTF--
 *** Test substituting argument 2 with array values ***
 
-Warning: getservbyname() expects parameter 2 to be string, array given in %s on line %d
+Warning: getservbyname() expects parameter 2 to be a valid path, array given in %s on line %d
 NULL
 
-Warning: getservbyname() expects parameter 2 to be string, array given in %s on line %d
+Warning: getservbyname() expects parameter 2 to be a valid path, array given in %s on line %d
 NULL
 
-Warning: getservbyname() expects parameter 2 to be string, array given in %s on line %d
+Warning: getservbyname() expects parameter 2 to be a valid path, array given in %s on line %d
 NULL
 
-Warning: getservbyname() expects parameter 2 to be string, array given in %s on line %d
+Warning: getservbyname() expects parameter 2 to be a valid path, array given in %s on line %d
 NULL
diff --git a/ext/standard/tests/network/ghsa-www2-q4fc-65wf.phpt b/ext/standard/tests/network/ghsa-www2-q4fc-65wf.phpt
new file mode 100644
index 00000000..0f7cbe1e
--- /dev/null
+++ b/ext/standard/tests/network/ghsa-www2-q4fc-65wf.phpt
@@ -0,0 +1,43 @@
+--TEST--
+GHSA-www2-q4fc-65wf
+--DESCRIPTION--
+This is a ZPP test but *keep* this as it is security-sensitive!
+--FILE--
+<?php
+var_dump(dns_check_record("\0"));
+var_dump(dns_get_mx("\0", $out));
+var_dump(dns_get_record("\0"));
+var_dump(getprotobyname("\0"));
+var_dump(getservbyname("\0", "tcp"));
+var_dump(getservbyname("x", "tcp\0"));
+var_dump(getservbyport(0, "tcp\0"));
+var_dump(inet_pton("\0"));
+var_dump(ip2long("\0"));
+?>
+--EXPECTF--
+Warning: dns_check_record() expects parameter 1 to be a valid path, string given in %s on line %d
+NULL
+
+Warning: dns_get_mx() expects parameter 1 to be a valid path, string given in %s on line %d
+NULL
+
+Warning: dns_get_record() expects parameter 1 to be a valid path, string given in %s on line %d
+NULL
+
+Warning: getprotobyname() expects parameter 1 to be a valid path, string given in %s on line %d
+NULL
+
+Warning: getservbyname() expects parameter 1 to be a valid path, string given in %s on line %d
+NULL
+
+Warning: getservbyname() expects parameter 2 to be a valid path, string given in %s on line %d
+NULL
+
+Warning: getservbyport() expects parameter 2 to be a valid path, string given in %s on line %d
+NULL
+
+Warning: inet_pton() expects parameter 1 to be a valid path, string given in %s on line %d
+bool(false)
+
+Warning: ip2long() expects parameter 1 to be a valid path, string given in %s on line %d
+NULL
diff --git a/ext/standard/tests/network/ip.phpt b/ext/standard/tests/network/ip.phpt
index 03329802..8e36199c 100644
--- a/ext/standard/tests/network/ip.phpt
+++ b/ext/standard/tests/network/ip.phpt
@@ -54,7 +54,7 @@ bool(false)
 bool(false)
 int(1869573999)
 
-Warning: ip2long() expects parameter 1 to be string, array given in %sip.php on line %d
+Warning: ip2long() expects parameter 1 to be a valid path, array given in %sip.php on line %d
 NULL
 
 Warning: long2ip() expects exactly 1 parameter, 0 given in %sip.php on line %d
diff --git a/ext/standard/tests/network/ip2long_variation1.phpt b/ext/standard/tests/network/ip2long_variation1.phpt
index fa741093..b305ae41 100644
--- a/ext/standard/tests/network/ip2long_variation1.phpt
+++ b/ext/standard/tests/network/ip2long_variation1.phpt
@@ -141,19 +141,19 @@ bool(false)
 bool(false)
 
 --empty array--
-Error: 2 - ip2long() expects parameter 1 to be string, array given, %s(%d)
+Error: 2 - ip2long() expects parameter 1 to be a valid path, array given, %s(%d)
 NULL
 
 --int indexed array--
-Error: 2 - ip2long() expects parameter 1 to be string, array given, %s(%d)
+Error: 2 - ip2long() expects parameter 1 to be a valid path, array given, %s(%d)
 NULL
 
 --associative array--
-Error: 2 - ip2long() expects parameter 1 to be string, array given, %s(%d)
+Error: 2 - ip2long() expects parameter 1 to be a valid path, array given, %s(%d)
 NULL
 
 --nested arrays--
-Error: 2 - ip2long() expects parameter 1 to be string, array given, %s(%d)
+Error: 2 - ip2long() expects parameter 1 to be a valid path, array given, %s(%d)
 NULL
 
 --uppercase NULL--
@@ -184,7 +184,7 @@ bool(false)
 bool(false)
 
 --instance of classWithoutToString--
-Error: 2 - ip2long() expects parameter 1 to be string, object given, %s(%d)
+Error: 2 - ip2long() expects parameter 1 to be a valid path, object given, %s(%d)
 NULL
 
 --undefined var--
@@ -194,6 +194,6 @@ bool(false)
 bool(false)
 
 --resource--
-Error: 2 - ip2long() expects parameter 1 to be string, resource given, %s(%d)
+Error: 2 - ip2long() expects parameter 1 to be a valid path, resource given, %s(%d)
 NULL
 ===DONE===
diff --git a/ext/standard/tests/network/ip_x86_64.phpt b/ext/standard/tests/network/ip_x86_64.phpt
index 7b93815f..6896063d 100644
--- a/ext/standard/tests/network/ip_x86_64.phpt
+++ b/ext/standard/tests/network/ip_x86_64.phpt
@@ -54,7 +54,7 @@ bool(false)
 bool(false)
 int(1869573999)
 
-Warning: ip2long() expects parameter 1 to be string, array given in %sip_x86_64.php on line %d
+Warning: ip2long() expects parameter 1 to be a valid path, array given in %sip_x86_64.php on line %d
 NULL
 
 Warning: long2ip() expects exactly 1 parameter, 0 given in %sip_x86_64.php on line %d
-- 
2.52.0

